pipelinedp4j/main/com/google/privacy/differentialprivacy/pipelinedp4j/spark/SparkTable.kt

```diff
@@ -0,0 +1,155 @@
+package com.google.privacy.differentialprivacy.pipelinedp4j.spark
+
+import com.google.privacy.differentialprivacy.pipelinedp4j.core.Encoder
+import com.google.privacy.differentialprivacy.pipelinedp4j.core.FrameworkCollection
+import com.google.privacy.differentialprivacy.pipelinedp4j.core.FrameworkTable
+import com.google.privacy.differentialprivacy.pipelinedp4j.local.LocalCollection
+import org.apache.spark.api.java.function.FlatMapFunction
+import org.apache.spark.api.java.function.MapFunction
+import org.apache.spark.api.java.function.MapGroupsFunction
+import org.apache.spark.api.java.function.ReduceFunction
+import org.apache.spark.sql.Dataset
+import org.apache.spark.sql.Encoders
+import org.apache.spark.sql.Row
+import scala.Tuple2
+
+class SparkTable<K, V>(val data: Dataset<Tuple2<K, V>>,
+                       val keyEncoder: org.apache.spark.sql.Encoder<K>,
+                       val valueEncoder: org.apache.spark.sql.Encoder<V>
+): FrameworkTable<K, V> {
+
+    override val keysEncoder = SparkEncoder(keyEncoder)
+    override val valuesEncoder = SparkEncoder(valueEncoder)
+    val keyValueEncoder = Encoders.tuple(keyEncoder, valueEncoder)
+    override fun <R> map(stageName: String, outputType: Encoder<R>, mapFn: (K, V) -> R): SparkCollection<R> {
+        val outputEncoder = (outputType as SparkEncoder<R>).encoder
+        val transformedData = data.map(MapFunction { kv: Tuple2<K, V> -> mapFn(kv._1, kv._2) }, outputEncoder)
+        return SparkCollection(transformedData)
+    }
+
+    override fun groupAndCombineValues(stageName: String, combFn: (V, V) -> V): SparkTable<K, V> {
+        val dataset = data
+            .groupByKey(MapFunction { kv: Tuple2<K, V> -> kv._1}, keyEncoder)
+            .reduceGroups(ReduceFunction { t1 : Tuple2<K, V>, t2: Tuple2<K, V> ->  Tuple2(t1._1, combFn(t1._2, t2._2))})
+            .map(MapFunction {it._2}, keyValueEncoder)
+        return SparkTable(dataset, keyEncoder, valueEncoder)
+    }
+    override fun groupByKey(stageName: String): SparkTable<K, Iterable<V>> {
+        val itrEncoder = Encoders.kryo(Iterable::class.java) as org.apache.spark.sql.Encoder<Iterable<V>>
+        val dats = data
+            .groupByKey(MapFunction { kv: Tuple2<K, V> -> kv._1}, keyEncoder)
+            .mapValues(MapFunction { kv: Tuple2<K, V> -> kv._2}, valueEncoder)
+            .mapGroups(MapGroupsFunction { k: K, v: MutableIterator<V> -> Tuple2(k, v.asSequence().toList()) },
```

Comment:
interesting, `MutableIterator` is Kotlin class, right? I thought we will have here Scala iterator. 

`v.asSequence().toList()` I would do it differently: `Iterable { v }`. This way we won't require the sequence to fit into RAM.